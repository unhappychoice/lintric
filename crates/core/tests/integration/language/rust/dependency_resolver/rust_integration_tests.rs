use insta::assert_snapshot;
use lintric_core::analyze_code;
use std::fs;

macro_rules! test_dependency_resolver_analysis {
    ($test_name:ident, $fixture_name:literal, $ir_snapshot:literal) => {
        #[test]
        fn $test_name() {
            let file_path = format!(
                "tests/integration/language/rust/dependency_resolver/fixtures/{}.rs",
                $fixture_name
            );
            let (ir, _result) = analyze_code(file_path.clone()).unwrap();

            let source_code = fs::read_to_string(&file_path).unwrap();
            let ast_result = lintric_core::get_s_expression_from_content(
                source_code.clone(),
                lintric_core::Language::Rust,
            );
            let ast_string = ast_result.unwrap_or_else(|e| format!("AST Error: {}", e));

            let ir_snapshot_content = format!(
                "Source Code:\n{}\n\nAST:\n{}\n\nIR:\n{:#?}",
                source_code, ast_string, ir
            );

            assert_snapshot!($ir_snapshot, ir_snapshot_content);

            // Verify dependencies are detected
            assert!(!ir.dependencies.is_empty(), "Should have dependencies");
        }
    };
}

test_dependency_resolver_analysis!(
    test_basic_symbol_resolution,
    "basic_symbol_resolution",
    "dependency_resolver_basic_symbol_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_method_resolution,
    "method_resolution",
    "dependency_resolver_method_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_generic_resolution,
    "generic_resolution",
    "dependency_resolver_generic_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_complex_trait_hierarchy,
    "complex_trait_hierarchy",
    "dependency_resolver_complex_trait_hierarchy_ir"
);

test_dependency_resolver_analysis!(
    test_scope_resolution,
    "scope_resolution",
    "dependency_resolver_scope_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_module_resolution,
    "module_resolution",
    "dependency_resolver_module_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_forward_references,
    "forward_references",
    "dependency_resolver_forward_references_ir"
);

test_dependency_resolver_analysis!(
    test_closure_dependencies,
    "closure_dependencies",
    "dependency_resolver_closure_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_advanced_generics,
    "advanced_generics",
    "dependency_resolver_advanced_generics_ir"
);

test_dependency_resolver_analysis!(
    test_macro_dependencies,
    "macro_dependencies",
    "dependency_resolver_macro_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_async_dependencies,
    "async_dependencies",
    "dependency_resolver_async_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_pattern_matching,
    "pattern_matching",
    "dependency_resolver_pattern_matching_ir"
);

#[test]
fn test_specific_dependency_types() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/method_resolution.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Check that we have some dependencies
    assert!(!ir.dependencies.is_empty(), "Should have some dependencies");
}

#[test]
fn test_trait_method_dependencies() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/complex_trait_hierarchy.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Verify trait method dependencies are resolved
    let trait_methods: Vec<_> = ir
        .dependencies
        .iter()
        .filter(|d| d.symbol == "name" || d.symbol == "speak" || d.symbol == "fur_color")
        .collect();

    assert!(
        !trait_methods.is_empty(),
        "Should resolve trait method dependencies"
    );
}

#[test]
fn test_generic_type_dependencies() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/generic_resolution.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Check for generic-related dependencies
    let generic_deps: Vec<_> = ir
        .dependencies
        .iter()
        .filter(|d| d.symbol.contains("Container") || d.symbol == "process")
        .collect();

    assert!(
        !generic_deps.is_empty(),
        "Should resolve generic type dependencies"
    );
}
