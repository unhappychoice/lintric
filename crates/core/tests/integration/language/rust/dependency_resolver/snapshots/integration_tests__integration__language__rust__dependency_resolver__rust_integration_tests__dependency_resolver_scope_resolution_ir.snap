---
source: crates/core/tests/integration/language/rust/dependency_resolver/rust_integration_tests.rs
expression: ir_snapshot_content
---
Source Code:
fn outer() {
    let outer_var = 42;
    
    fn inner() {
        println!("{}", outer_var);
    }
    
    {
        let block_var = 20;
        println!("{} {}", outer_var, block_var);
    }
    
    inner();
}

fn main() {
    let x = 10;
    
    {
        let y = 20;
        println!("{} {}", x, y);
    }
    
    println!("{}", x);
}

AST:
(source_file
  (function_item
    name: (identifier "outer")
    parameters: (parameters)
    body: (block
      (let_declaration
        pattern: (identifier "outer_var")
        value: (integer_literal "42")
      )
      (function_item
        name: (identifier "inner")
        parameters: (parameters)
        body: (block
          (expression_statement
            (macro_invocation
              macro: (identifier "println")
              (token_tree
                (string_literal ""{}""
                  (string_content)
                )
                (identifier "outer_var")
              )
            )
          )
        )
      )
      (expression_statement
        (block
          (let_declaration
            pattern: (identifier "block_var")
            value: (integer_literal "20")
          )
          (expression_statement
            (macro_invocation
              macro: (identifier "println")
              (token_tree
                (string_literal ""{} {}""
                  (string_content)
                )
                (identifier "outer_var")
                (identifier "block_var")
              )
            )
          )
        )
      )
      (expression_statement
        (call_expression
          function: (identifier "inner")
          arguments: (arguments)
        )
      )
    )
  )
  (function_item
    name: (identifier "main")
    parameters: (parameters)
    body: (block
      (let_declaration
        pattern: (identifier "x")
        value: (integer_literal "10")
      )
      (expression_statement
        (block
          (let_declaration
            pattern: (identifier "y")
            value: (integer_literal "20")
          )
          (expression_statement
            (macro_invocation
              macro: (identifier "println")
              (token_tree
                (string_literal ""{} {}""
                  (string_content)
                )
                (identifier "x")
                (identifier "y")
              )
            )
          )
        )
      )
      (expression_statement
        (macro_invocation
          macro: (identifier "println")
          (token_tree
            (string_literal ""{}""
              (string_content)
            )
            (identifier "x")
          )
        )
      )
    )
  )
)

IR:
IntermediateRepresentation {
    file_path: "tests/integration/language/rust/dependency_resolver/fixtures/scope_resolution.rs",
    definitions: [
        Definition {
            name: "outer",
            position: Position {
                start_line: 1,
                start_column: 4,
                end_line: 1,
                end_column: 9,
            },
            definition_type: FunctionDefinition,
        },
        Definition {
            name: "outer_var",
            position: Position {
                start_line: 2,
                start_column: 9,
                end_line: 2,
                end_column: 18,
            },
            definition_type: VariableDefinition,
        },
        Definition {
            name: "inner",
            position: Position {
                start_line: 4,
                start_column: 8,
                end_line: 4,
                end_column: 13,
            },
            definition_type: FunctionDefinition,
        },
        Definition {
            name: "block_var",
            position: Position {
                start_line: 9,
                start_column: 13,
                end_line: 9,
                end_column: 22,
            },
            definition_type: VariableDefinition,
        },
        Definition {
            name: "main",
            position: Position {
                start_line: 16,
                start_column: 4,
                end_line: 16,
                end_column: 8,
            },
            definition_type: FunctionDefinition,
        },
        Definition {
            name: "x",
            position: Position {
                start_line: 17,
                start_column: 9,
                end_line: 17,
                end_column: 10,
            },
            definition_type: VariableDefinition,
        },
        Definition {
            name: "y",
            position: Position {
                start_line: 20,
                start_column: 13,
                end_line: 20,
                end_column: 14,
            },
            definition_type: VariableDefinition,
        },
    ],
    dependencies: [
        Dependency {
            source_line: 5,
            target_line: 2,
            symbol: "outer_var",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:5:24",
            ),
        },
        Dependency {
            source_line: 10,
            target_line: 2,
            symbol: "outer_var",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:10:27",
            ),
        },
        Dependency {
            source_line: 10,
            target_line: 9,
            symbol: "block_var",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:10:38",
            ),
        },
        Dependency {
            source_line: 13,
            target_line: 4,
            symbol: "inner",
            dependency_type: FunctionCall,
            context: Some(
                "CallExpression:13:5",
            ),
        },
        Dependency {
            source_line: 21,
            target_line: 17,
            symbol: "x",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:21:27",
            ),
        },
        Dependency {
            source_line: 21,
            target_line: 20,
            symbol: "y",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:21:30",
            ),
        },
        Dependency {
            source_line: 24,
            target_line: 17,
            symbol: "x",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:24:20",
            ),
        },
    ],
    usage: [
        Usage {
            name: "println",
            kind: Identifier,
            position: Position {
                start_line: 5,
                start_column: 9,
                end_line: 5,
                end_column: 16,
            },
        },
        Usage {
            name: "outer_var",
            kind: Identifier,
            position: Position {
                start_line: 5,
                start_column: 24,
                end_line: 5,
                end_column: 33,
            },
        },
        Usage {
            name: "println",
            kind: Identifier,
            position: Position {
                start_line: 10,
                start_column: 9,
                end_line: 10,
                end_column: 16,
            },
        },
        Usage {
            name: "outer_var",
            kind: Identifier,
            position: Position {
                start_line: 10,
                start_column: 27,
                end_line: 10,
                end_column: 36,
            },
        },
        Usage {
            name: "block_var",
            kind: Identifier,
            position: Position {
                start_line: 10,
                start_column: 38,
                end_line: 10,
                end_column: 47,
            },
        },
        Usage {
            name: "inner",
            kind: CallExpression,
            position: Position {
                start_line: 13,
                start_column: 5,
                end_line: 13,
                end_column: 12,
            },
        },
        Usage {
            name: "println",
            kind: Identifier,
            position: Position {
                start_line: 21,
                start_column: 9,
                end_line: 21,
                end_column: 16,
            },
        },
        Usage {
            name: "x",
            kind: Identifier,
            position: Position {
                start_line: 21,
                start_column: 27,
                end_line: 21,
                end_column: 28,
            },
        },
        Usage {
            name: "y",
            kind: Identifier,
            position: Position {
                start_line: 21,
                start_column: 30,
                end_line: 21,
                end_column: 31,
            },
        },
        Usage {
            name: "println",
            kind: Identifier,
            position: Position {
                start_line: 24,
                start_column: 5,
                end_line: 24,
                end_column: 12,
            },
        },
        Usage {
            name: "x",
            kind: Identifier,
            position: Position {
                start_line: 24,
                start_column: 20,
                end_line: 24,
                end_column: 21,
            },
        },
    ],
    analysis_metadata: AnalysisMetadata {
        language: "Rust",
        total_lines: 25,
        analysis_timestamp: "now",
        lintric_version: "0.1.0",
    },
}
